src/lib.rs

```diff
@@ -2094,31 +2143,22 @@ extern "C" fn fn_decrypt_digest_update(
     }
     let rstate = global_rlock!(STATE);
     let mut session = res_or_ret!(rstate.get_session_mut(s_handle));
+
     res_or_ret!(session.check_op::<dyn Digest>());
-    let operation = res_or_ret!(session.get_operation::<dyn Decryption>());
-    let elen = cast_or_ret!(usize from encrypted_part_len => CKR_ARGUMENTS_BAD);
+
+    res_or_ret!(internal_decrypt_update(
+        &mut session,
+        encrypted_part,
+        encrypted_part_len,
+        part,
+        pul_part_len,
+    ));
     if part.is_null() {
-        let decryption_len = cast_or_ret!(
-            CK_ULONG from res_or_ret!(operation.decryption_len(elen, false))
-        );
-        unsafe {
-            *pul_part_len = decryption_len;
-        }
         return CKR_OK;
     }
-    let enc: &[u8] =
-        unsafe { std::slice::from_raw_parts(encrypted_part, elen) };
-    let pplen = unsafe { *pul_part_len as CK_ULONG };
-    let plen = cast_or_ret!(usize from pplen => CKR_ARGUMENTS_BAD);
-    let dpart: &mut [u8] =
-        unsafe { std::slice::from_raw_parts_mut(part, plen) };
-    let outlen = res_or_ret!(operation.decrypt_update(enc, dpart));
-    let retlen = cast_or_ret!(CK_ULONG from outlen);
-    unsafe { *pul_part_len = retlen };
-    let retpart: &mut [u8] =
-        unsafe { std::slice::from_raw_parts_mut(part, outlen) };
-    let operation = res_or_ret!(session.get_operation::<dyn Digest>());
-    ret_to_rv!(operation.digest_update(retpart))
+
+    let part_len: CK_ULONG = unsafe { *pul_part_len };
+    ret_to_rv!(internal_digest_update(&mut session, part, part_len,))
```

Comment:
trailing comma in inline argument list is ok from both functional point of view as well as rust-fmt, but distracting to read. Would be good to get rid of that. The same in `fn_digest_encrypt_update()` and other dual functions. Is there some linter that could enforce these? I saw clippy in the rust-openssl, but did not look into that much further ...

Reply:
I with rust-format would better handle these, it is its doing ...
I write freestyle then call `make fix-format` before pushing ...